File: Go/100. Same Tree.go
package main

/*100. Same Tree
给定两个二叉树，编写一个函数来检验它们是否相同。
如果两个树在结构上相同，并且节点具有相同的值，则认为它们是相同的。

示例:
输入:       1         1
          / \       / \
         2   3     2   3

        [1,2,3],   [1,2,3]
输出: true
*/
// import "fmt"

//TreeNode Definition for a binary tree node.
// type TreeNode struct {
// 	Val   int
// 	Left  *TreeNode
// 	Right *TreeNode
// }

func isSameTree(p *TreeNode, q *TreeNode) bool {
	if p == nil && q == nil {
		return true
	} else if p == nil || q == nil {
		return false
	} else {
		if p.Val != q.Val {
			return false
		}
	}
	return isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
}

// func main() {
// 	var p, q *TreeNode
// 	p, q = nil, nil
// 	print(isSameTree(q, p))

// }

File: Rust/src/100. Same Tree.rs
use std::rc::Rc;
use std::cell::RefCell;
/*

*/
//Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
  pub fn new(val: i32) -> Self {
      TreeNode {
      val,
      left: None,
      right: None
    }
}
}
impl Solution {
    pub fn is_same_tree(p: Option<Rc<RefCell<TreeNode>>>, q: Option<Rc<RefCell<TreeNode>>>) -> bool {
        match(p,q){
            (None,None)=>true,
            (None,_)|(_,None)=>false,
            (Some(pr),Some(qr))=>{
                let (pr,qr)=(pr.as_ref(),qr.as_ref());
                let (pr,qr)=(pr.borrow(),qr.borrow());
                if pr.val==qr.val{
                    Self::is_same_tree(pr.left.clone(),qr.left.clone())&&Self::is_same_tree(pr.right.clone(),qr.right.clone())
                }else{
                    false
                }
            }
        }
    }
}
struct Solution{
    
}

impl Solution{
}

fn main(){

}